modules/sensor_interface.py

```diff
@@ -30,41 +32,96 @@ def _collect_data(self) -> None:
         """Continuously collect data from sensors."""
         while self.is_collecting:
             try:
-                # Placeholder for actual sensor reading logic
-                # TODO: Implement actual sensor reading here
-                sensor_reading = self._read_distance_sensor()
+                # Real-time sensor readings
+                ultrasonic_reading = self._read_ultrasonic_sensor()
+                lidar_reading = self._read_lidar_sensor()
                 
                 with self._lock:
-                    self.sensor_data.update(sensor_reading)
+                    self.sensor_data.update({
+                        'ultrasonic': ultrasonic_reading,
+                        'lidar': lidar_reading
+                    })
                 
-                # Add to queue for processing
-                if not self.data_queue.full():
-                    self.data_queue.put(sensor_reading)
+                print(f"Ultrasonic Reading: {ultrasonic_reading}")  # Debug log
+                print(f"Lidar Reading: {lidar_reading}")  # Debug log
                 
-                time.sleep(0.1)  # Adjust sampling rate as needed
+                time.sleep(0.5)  # Adjust sampling rate as needed
                 
             except Exception as e:
                 print(f"Error collecting sensor data: {e}")
-                time.sleep(1)  # Wait before retrying
+                time.sleep(0.5)  # Wait before retrying
 
-    def _read_distance_sensor(self) -> Dict[str, float]:
-        """Read data from distance sensor.
+    def _read_ultrasonic_sensor(self) -> Dict[str, float]:
+        """Read data from ultrasonic sensor using GPIO"""
         
-        TODO: Implement actual sensor reading logic using GPIO
-        This is a placeholder that returns dummy data.
-        """
+        TRIG_PIN = 23  # Replace with your actual TRIG pin number
+        ECHO_PIN = 24  # Replace with your actual ECHO pin number
+
+        GPIO.setmode(GPIO.BCM)
+        GPIO.setup(TRIG_PIN, GPIO.OUT)
+        GPIO.setup(ECHO_PIN, GPIO.IN)
+
+        # Send a 10us pulse to trigger the sensor
+        GPIO.output(TRIG_PIN, True)
+        time.sleep(0.00001)
+        GPIO.output(TRIG_PIN, False)
+
+        # Measure the time for the echo to return
+        pulse_start = time.time()
+        pulse_end = time.time()
+
+        while GPIO.input(ECHO_PIN) == 0:
+            pulse_start = time.time()
+
+        while GPIO.input(ECHO_PIN) == 1:
+            pulse_end = time.time()
+
+        # Calculate distance in cm
+        pulse_duration = pulse_end - pulse_start
+        distance = pulse_duration * 17150  # Speed of sound: 34300 cm/s divided by 2
```

Comment:
```suggestion
        distance = (pulse_duration * 34300) / 2  # Speed of sound: 34300 cm/s divided by 2
```
Just a suggestion. This might keep the logic even if the comment is removed on the magic number.

Reply:
Updated logic and removed comment.